File: src/graph-and-tree/smallest-subtree-with-all-the-deepest-nodes.ts
export function subtreeWithAllDeepest(root: TreeNode | null): TreeNode | null {
  type LCA = [node: TreeNode | null, depth: number]

  const Impl = (node: TreeNode | null): LCA => {
    if (!node) {
      return [null, 0]
    }

    const left = Impl(node.left)
    const right = Impl(node.right)

    if (left[1] > right[1]) {
      return [left[0], left[1] + 1]
    }

    if (right[1] > left[1]) {
      return [right[0], right[1] + 1]
    }

    return [node, left[1] + 1]
  }

  return Impl(root)[0]
}

File: src/graph-and-tree/smallest-subtree-with-all-the-deepest-nodes.cpp
class Solution {
 public:
  pair<TreeNode*, int> f(TreeNode* root) {
    if (!root) {
      return {root, 0};
    }

    auto left = f(root->left);
    auto right = f(root->right);

    if (left.second > right.second) {
      return {left.first, left.second + 1};
    }
    if (left.second < right.second) {
      return {right.first, right.second + 1};
    }
    return {root, left.second + 1};
  }

  TreeNode* subtreeWithAllDeepest(TreeNode* root) { return f(root).first; }
};